File: components/autocomplete/android/CiataAutocomplete.kt
package org.ciata.ds.autocomplete

import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable

/** Implementação experimental do CMP-0022 Autocomplete para Jetpack Compose. */
@Composable
fun CiataAutocomplete(
    label: String,
    value: String,
    options: List<String>,
    expanded: Boolean,
    onValueChange: (String) -> Unit,
    onExpandedChange: (Boolean) -> Unit,
    onSelect: (String) -> Unit,
) {
    val normalizedLabel = label.trim()
    val normalizedOptions = options.map(String::trim)

    require(normalizedLabel.isNotEmpty()) { "label não pode ser vazio." }
    require(normalizedOptions.none { it.isEmpty() }) { "options não pode conter rótulos vazios." }
    require(normalizedOptions.distinct().size == normalizedOptions.size) { "options não pode conter rótulos duplicados." }

    val effectiveExpanded = expanded && value.isNotBlank() && normalizedOptions.isNotEmpty()

    OutlinedTextField(
        value = value,
        onValueChange = {
            onValueChange(it)
            onExpandedChange(it.isNotBlank() && normalizedOptions.isNotEmpty())
        },
        singleLine = true,
        label = { Text(normalizedLabel) },
    )
    DropdownMenu(
        expanded = effectiveExpanded,
        onDismissRequest = { onExpandedChange(false) },
    ) {
        normalizedOptions.forEach { option ->
            DropdownMenuItem(
                text = { Text(option) },
                onClick = {
                    onSelect(option)
                    onExpandedChange(false)
                },
            )
        }
    }
}

File: components/autocomplete/ios/CiataAutocomplete.swift
import SwiftUI

/// Implementação experimental do CMP-0022 Autocomplete para SwiftUI.
public struct CiataAutocomplete: View {
    private let label: String
    private let options: [String]
    @Binding private var value: String
    private let onSelect: (String) -> Void
    @State private var isExpanded = false

    public init(
        label: String,
        value: Binding<String>,
        options: [String],
        onSelect: @escaping (String) -> Void
    ) {
        let trimmed = label.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedOptions = options.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        precondition(!trimmed.isEmpty, "label não pode ser vazio")
        precondition(!normalizedOptions.contains { $0.isEmpty }, "options não pode conter rótulos vazios")
        self.label = trimmed
        self._value = value
        self.options = normalizedOptions
        self.onSelect = onSelect
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            TextField(label, text: $value)
                .onChange(of: value) { _, newValue in
                    isExpanded = !newValue.isEmpty && !options.isEmpty
                }
            if isExpanded {
                ForEach(Array(options.enumerated()), id: \.offset) { _, option in
                    Button(option) {
                        value = option
                        onSelect(option)
                        isExpanded = false
                    }
                }
            }
        }
    }
}
